edge-hub/src/Microsoft.Azure.Devices.Routing.Core/endpoints/StoringAsyncEndpointExecutor.cs

```diff
@@ -23,51 +25,50 @@ public class StoringAsyncEndpointExecutor : IEndpointExecutor
         readonly IMessageStore messageStore;
         readonly Task sendMessageTask;
         readonly AsyncManualResetEvent hasMessagesInQueue = new AsyncManualResetEvent(true);
-        readonly ICheckpointer checkpointer;
-        readonly AtomicReference<IList<uint>> priorities = new AtomicReference<IList<uint>>(new List<uint>());
         readonly AsyncEndpointExecutorOptions options;
-        readonly EndpointExecutorFsm machine;
         readonly CancellationTokenSource cts = new CancellationTokenSource();
+        readonly ICheckpointerFactory checkpointerFactory;
+        readonly EndpointExecutorConfig config;
+        AtomicReference<OrderedDictionary> prioritiesToFsms;
```

Comment:
Just curious, can we use ConcurrentDictionary instead of AtomicReference<OrderedDictionary>?  @varunpuranik would you give some context about AtomicReference?

Reply:
We've used an `AtomicReference` and an immutable dictionary together before. This is because the immutable dictionary is updated by making a new version. This pairs nicely with AtomicReferences.

I think a concurrent dictionary is an issue because the keys need to be sorted?

Reply:
I am not sure that concurrent dictionary needed to be sorted; but actually its AddOrUpdate and GetOrAdd method doesn't execute under locks and need to be taken care by yourself.

Reply:
yes, the keys need to be sorted, and ConcurrentDictionary doesn't allow that.

Using AtomicReference here enforces a snapshot mechanism, so that the writer updating the dictionary doesn't mess up the reader.

---
In reply to: [392285442](https://github.com/Azure/iotedge/pull/2670#discussion_r392285442) [](ancestors = 392285442)

Reply:
the message pump needs the dictionary to be sorted, so that it can iterate over messages in ascending priority order.

---
In reply to: [392302436](https://github.com/Azure/iotedge/pull/2670#discussion_r392302436) [](ancestors = 392302436)